crypto/fipsmodule/hmac/hmac.c

```diff
@@ -120,119 +175,114 @@ void HMAC_CTX_free(HMAC_CTX *ctx) {
 
 int HMAC_Init_ex(HMAC_CTX *ctx, const void *key, size_t key_len,
                  const EVP_MD *md, ENGINE *impl) {
-  // We have to avoid the underlying SHA services updating the indicator
-  // state, so we lock the state here.
-  FIPS_service_indicator_lock_state();
-  int ret = 0;
-  if (md == NULL) {
-    md = ctx->md;
+  // TODO: Figure out how we want to handle engines in this case.
+  assert(impl == NULL);
+
+  if (md && ctx->md != md) {
+    ctx->methods = GetInPlaceMethods(md);
+    if (ctx->methods == NULL) {
+      // Unsupported md
+      return 0;
+    }
+    ctx->md = md;
+  } else if (ctx->initialized != 1) {
+    return 0;
   }
 
-  // If either |key| is non-NULL or |md| has changed, initialize with a new key
-  // rather than rewinding the previous one.
-  //
-  // TODO(davidben,eroman): Passing the previous |md| with a NULL |key| is
-  // ambiguous between using the empty key and reusing the previous key. There
-  // exist callers which intend the latter, but the former is an awkward edge
-  // case. Fix to API to avoid this.
-  if (md != ctx->md || key != NULL) {
-    uint8_t pad[EVP_MAX_MD_BLOCK_SIZE];
-    uint8_t key_block[EVP_MAX_MD_BLOCK_SIZE];
-    unsigned key_block_len;
-
-    size_t block_size = EVP_MD_block_size(md);
-    assert(block_size <= sizeof(key_block));
+  // At this point we know we have valid methods and a context allocated.
+  const HmacMethods *methods = ctx->methods;
+  size_t block_size = EVP_MD_block_size(methods->evp_md);
+  assert(block_size % 8 == 0);
+  assert(block_size <= EVP_MAX_MD_BLOCK_SIZE);
+
+  FIPS_service_indicator_lock_state();
+  int result = 0;
+  if (key != NULL || ctx->initialized != 1) {
+    uint64_t pad[EVP_MAX_MD_BLOCK_SIZE] = {0};
+    uint64_t key_block[EVP_MAX_MD_BLOCK_SIZE] = {0};
     if (block_size < key_len) {
-      // Long keys are hashed.
-      if (!EVP_DigestInit_ex(&ctx->md_ctx, md, impl) ||
-          !EVP_DigestUpdate(&ctx->md_ctx, key, key_len) ||
-          !EVP_DigestFinal_ex(&ctx->md_ctx, key_block, &key_block_len)) {
+      if (!methods->init(&ctx->md_ctx) ||
+          !methods->update(&ctx->md_ctx, key, key_len) ||
+          !methods->digest((uint8_t *) key_block, &ctx->md_ctx)) {
         goto end;
       }
     } else {
-      assert(key_len <= sizeof(key_block));
       OPENSSL_memcpy(key_block, key, key_len);
-      key_block_len = (unsigned)key_len;
     }
-    // Keys are then padded with zeros.
-    if (key_block_len != EVP_MAX_MD_BLOCK_SIZE) {
-      OPENSSL_memset(&key_block[key_block_len], 0, sizeof(key_block) - key_block_len);
+    for (size_t i = 0; i < block_size / 8; i++) {
```

Comment:
The previous version of this function seems to allow key_len to not be divisible by 8 when padding with 0s. However, after that, I think both the previous and this version, due to XORing the ipad and opad with 0, will have those patterns show in key_block. Are we relying on the actual implementation to ignore those extra bytes?

Reply:
I don't understand this comment. Neither the old nor new implementation require that the key_len be divisible by 8. The new implementation requires that the `block_size` be divisible by 8. All currently supported hash functions meet this requirement.